ddl/backfilling_operator.go

```diff
@@ -0,0 +1,511 @@
+	if err != nil {
+		idxResult.Err = err
+		sender(idxResult)
+	}
+}
+
+func (w *tableScanWorker) getChunk() *chunk.Chunk {
+	chk := <-w.srcChkPool
+	newCap := copReadBatchSize()
+	if chk.Capacity() != newCap {
+		chk = chunk.NewChunkWithCapacity(w.copCtx.fieldTps, newCap)
+	}
+	chk.Reset()
+	return chk
+}
+
+func (w *tableScanWorker) recycleChunk(chk *chunk.Chunk) {
+	w.srcChkPool <- chk
+}
+
+// IndexWriteResult contains the result of writing index records to ingest engine.
+type IndexWriteResult struct {
+	ID      int
+	Added   int
+	Scanned int
+	Total   int
+	Next    kv.Key
+	Err     error
+}
+
+// IndexIngestOperator writes index records to ingest engine.
+type IndexIngestOperator struct {
+	*operator.AsyncOperator[IndexRecordChunk, IndexWriteResult]
+}
+
+// NewIndexIngestOperator creates a new IndexIngestOperator.
+func NewIndexIngestOperator(
+	ctx context.Context,
+	copCtx *copContext,
+	sessPool opSessPool,
+	tbl table.PhysicalTable,
+	index table.Index,
+	engine ingest.Engine,
+	srcChunkPool chan *chunk.Chunk,
+	concurrency int,
+) *IndexIngestOperator {
+	var writerIDAlloc atomic.Int32
+	pool := workerpool.NewWorkerPool(
+		"indexIngestOperator",
+		util.DDL,
+		concurrency,
+		func() workerpool.Worker[IndexRecordChunk, IndexWriteResult] {
+			writerID := int(writerIDAlloc.Add(1))
+			writer, err := engine.CreateWriter(writerID, index.Meta().Unique)
+			if err != nil {
+				logutil.Logger(ctx).Error("create index ingest worker failed", zap.Error(err))
+				return nil
+			}
+			return &indexIngestWorker{
+				ctx:          ctx,
+				tbl:          tbl,
+				index:        index,
+				copCtx:       copCtx,
+				se:           nil,
+				sessPool:     sessPool,
+				writer:       writer,
+				srcChunkPool: srcChunkPool,
+			}
+		})
+	return &IndexIngestOperator{
+		AsyncOperator: operator.NewAsyncOperator[IndexRecordChunk, IndexWriteResult](ctx, pool),
+	}
+}
+
+type indexIngestWorker struct {
+	ctx context.Context
+
+	tbl   table.PhysicalTable
+	index table.Index
+
+	copCtx   *copContext
+	sessPool opSessPool
+	se       *session.Session
+
+	writer       ingest.Writer
+	srcChunkPool chan *chunk.Chunk
+}
+
+func (w *indexIngestWorker) HandleTask(rs IndexRecordChunk, sender func(IndexWriteResult)) {
+	result := IndexWriteResult{
+		ID:  rs.ID,
+		Err: rs.Err,
+	}
+	if w.se == nil {
+		sessCtx, err := w.sessPool.Get()
+		if err != nil {
+			result.Err = err
+			sender(result)
+			return
+		}
+		w.se = session.NewSession(sessCtx)
+	}
+	defer func() {
+		w.srcChunkPool <- rs.Chunk
+	}()
+	if result.Err != nil {
+		logutil.Logger(w.ctx).Error("encounter error when handle index chunk",
+			zap.Int("id", rs.ID), zap.Error(rs.Err))
+		sender(result)
+		return
+	}
+	count, nextKey, err := w.WriteLocal(&rs)
+	if err != nil {
+		result.Err = err
+		sender(result)
+		return
+	}
+	if count == 0 {
+		logutil.Logger(w.ctx).Info("finish a cop-request task", zap.Int("id", rs.ID))
+		sender(result)
+		return
+	}
+	result.Added = count
+	result.Scanned = count
```

Comment:
What's the different between scanned and added?

Reply:
No difference here.. let me remove one of them.